server/src/internalClusterTest/java/org/elasticsearch/cluster/metadata/MetadataUpdateSettingsServiceIT.java

```diff
@@ -0,0 +1,92 @@
+
+    public void testThatNonDynamicSettingChangesTakeEffect() throws Exception {
+        /*
+         * This test makes sure that when non-dynamic settings are updated that they actually take effect (as opposed to just being set
+         * in the cluster state).
+         */
+        createIndex("test", Settings.EMPTY);
+        MetadataUpdateSettingsService metadataUpdateSettingsService = internalCluster().getCurrentMasterNodeInstance(
+            MetadataUpdateSettingsService.class
+        );
+        UpdateSettingsClusterStateUpdateRequest request = new UpdateSettingsClusterStateUpdateRequest();
+        List<Index> indices = new ArrayList<>();
+        for (IndicesService indicesService : internalCluster().getInstances(IndicesService.class)) {
+            for (IndexService indexService : indicesService) {
+                indices.add(indexService.index());
+            }
+        }
+        request.indices(indices.toArray(Index.EMPTY_ARRAY));
+        request.settings(Settings.builder().put("index.codec", "CheapBastard").build());
+
+        // First make sure it fails if reopenShards is not set on the request:
+        AtomicBoolean expectedFailureOccurred = new AtomicBoolean(false);
+        metadataUpdateSettingsService.updateSettings(request, new ActionListener<>() {
+            @Override
+            public void onResponse(AcknowledgedResponse acknowledgedResponse) {
+                fail("Should have failed updating a non-dynamic setting without reopenShards set to true");
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                expectedFailureOccurred.set(true);
+            }
+        });
+        assertBusy(() -> assertThat(expectedFailureOccurred.get(), equalTo(true)));
+
+        // Now we set reopenShards and expect it to work:
+        request.reopenShards(true);
+        AtomicBoolean success = new AtomicBoolean(false);
+        metadataUpdateSettingsService.updateSettings(request, new ActionListener<>() {
+            @Override
+            public void onResponse(AcknowledgedResponse acknowledgedResponse) {
+                success.set(true);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                fail(e);
+            }
+        });
+        assertBusy(() -> assertThat(success.get(), equalTo(true)));
+
+        // Now we look into the IndexShard objects to make sure that the code was actually updated (vs just the setting):
+        for (IndicesService indicesService : internalCluster().getInstances(IndicesService.class)) {
+            for (IndexService indexService : indicesService) {
+                for (IndexShard indexShard : indexService) {
+                    final Engine engine = indexShard.getEngineOrNull();
+                    assertNotNull(engine);
+                    assertThat(engine.getEngineConfig().getCodec().getName(), equalTo("CheapBastard"));
```

Comment:
:grin: maybe let's avoid the expletive here tho

Reply:
Ha it is actually one of our options. I ran into this because I thought from our documentation that my first choice, "best-compression", was different than the default, but it is actually the default (so no change). My thinking was that this one is never going to be the default. :)